controllers/hardwareclassification_controller.go

```diff
@@ -0,0 +1,172 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-logr/logr"
+
+	hwcc "hardware-classification-controller/api/v1alpha1"
+
+	bmh "github.com/metal3-io/baremetal-operator/pkg/apis/metal3/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// HardwareClassificationReconciler reconciles a HardwareClassification object
+type HardwareClassificationReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// Reconcile reconcile function
+// +kubebuilder:rbac:groups=metal3.io.sigs.k8s.io,resources=hardwareclassifications,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=metal3.io.sigs.k8s.io,resources=hardwareclassifications/status,verbs=get;update;patch
+func (r *HardwareClassificationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	// Get HardwareClassificationController to get values for Namespace and ExpectedHardwareConfiguration
+	hardwareClassification := &hwcc.HardwareClassification{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, hardwareClassification); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Get ExpectedHardwareConfiguraton from hardwareClassification
+	extractedProfile := hardwareClassification.Spec.ExpectedHardwareConfiguration
+
+	// Checking Disk Count present or not if Disk individual size is present
+	if extractedProfile.Disk != nil {
+		if err := checkDiskCount(extractedProfile); err != nil {
+			r.Log.Error(nil, "Disk count error", "Error", err.Error())
```

Comment:
This error message will not be visible to an API user unless they are also watching the controller logs. Is there a place to put this in the status of the profile that is mis-configured? 

On the other hand, why is it an error? What if I only care that there is a disk with X GB and I don't care how many other disks there are?

Reply:
on top of this, if I specify a min / max size, does it apply to all the disks, or only to one of the disks, to ensure there is at list one of that size, or it could even apply to the sum of all the disks?

Reply:
@nehaalhat You will have to put above error in CRD HardwareClassificationStatus structure so that it will populate in controller logs.

Also remove dependancy of disk size and disk count.